runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java

```diff
@@ -120,15 +122,113 @@ public LayoutWorkflowManager removeLayoutServers(Set<String> endpoints)
+     * @return this.
+     */
+    public LayoutBuilder addSequencerServer(String endpoint) {
+        layout.getSequencers().add(endpoint);
+        return this;
+    }
+
+    /**
+     * Adds a new logunit server.
+     * The latest open segment (which has its end address as infinity) is now closed at address
+     * 'globalLogTail'.
+     * A new segment is opened for all future writes. The segment starts at 'globalLogTail' and
+     * ends at infinity. This new segment contains the new logunit server to be added in the
+     * specified stripe.
+     *
+     * @param stripeIndex        stripe index where new endpoint should be added.
+     * @param globalLogTail      Global log tail to split segment.
+     * @param newLogunitEndpoint Endpoint to be added.
+     * @return this.
+     */
+    public LayoutBuilder addLogunitServer(int stripeIndex, long globalLogTail,
+                                          String newLogunitEndpoint) {
+        List<LayoutSegment> layoutSegmentList = layout.getSegments();
+        LayoutSegment segmentToSplit = layoutSegmentList.remove(layoutSegmentList.size() - 1);
+
+        // Close the existing segment with the provided globalLogTail.
+        LayoutSegment closedSegment = new LayoutSegment(segmentToSplit.getReplicationMode(),
+                segmentToSplit.getStart(),
+                globalLogTail + 1,
+                segmentToSplit.getStripes());
+
+        List<String> logunitServerList = new ArrayList<>();
+        logunitServerList.addAll(segmentToSplit.getStripes()
+                .get(stripeIndex)
+                .getLogServers());
+        logunitServerList.add(newLogunitEndpoint);
+
+        // Create new stripe list with the new logunit server added.
+        LayoutStripe newStripe = new LayoutStripe(logunitServerList);
+        List<LayoutStripe> newStripeList = new ArrayList<>();
+        newStripeList.addAll(segmentToSplit.getStripes());
+        newStripeList.remove(stripeIndex);
+        newStripeList.add(stripeIndex, newStripe);
+
+        LayoutSegment openSegment = new LayoutSegment(segmentToSplit.getReplicationMode(),
+                globalLogTail + 1,
+                segmentToSplit.getEnd(),
+                newStripeList);
+        layoutSegmentList.add(layoutSegmentList.size(), closedSegment);
+        layoutSegmentList.add(layoutSegmentList.size(), openSegment);
+        return this;
+    }
+
+    /**
+     * Merges the specified segment and the segment before this.
+     *
+     * @param segmentIndex Segment to merge.
+     * @return this.
+     * @throws LayoutModificationException
+     */
+    public LayoutBuilder mergePreviousSegment(int segmentIndex)
+            throws LayoutModificationException {
+        if (segmentIndex < 1) {
+            throw new LayoutModificationException("No segments to merge.");
+        }
+
+        List<LayoutSegment> layoutSegmentList = layout.getSegments();
+        if (layoutSegmentList.get(segmentIndex).start
```

Comment:
This is wrong, consider the following example with two segments, [A, B1) [B2, C) , then this equality will check if B2 == A, which is wrong (i.e. you want to check if B1 == B2), you want to check if the two intervals are not continuous, so you want to check if  layoutSegmentList.get(segmentIndex).start != layoutSegmentList.get(segmentIndex - 1).end

Reply:
Fixed

Reply:
I think we should either restrict to a single split or handle multiple splits, i prefer the first option for now. For the later case, you need two segment indices to merge.